balancer.go

```diff
@@ -0,0 +1,329 @@
+package xlb
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"github.com/google/uuid"
+	"github.com/rs/zerolog"
+	"github.com/xdire/xlb/tlsutil"
+	"net"
+	"os"
+	"strings"
+	"sync"
+	"time"
+)
+
+type ServicePool interface {
+	// Identity How each ServicePool identified, CN match
+	Identity() string
+	// Port to listen for incoming traffic
+	Port() int
+	// RateQuota Rate of times per time.Duration
+	RateQuota() (int, time.Duration)
+	// Routes to route
+	Routes() []Route
+	// TLSData to service the frontend
+	TLSData() TLSData
+	// UnauthorizedAttempts How many unauthorized attempts before IP cache placement
+	UnauthorizedAttempts() int
+	// HealthCheckValidations Bring host back in routable healthy state after this amount of validations
+	HealthCheckValidations() int
+	// RouteTimeout general timeout for route to be connected or dropped
+	RouteTimeout() time.Duration
+}
+
+type Route interface {
+	// Path Stores path of the upstream
+	Path() string
+	// Active Provides information if route is active, in case of update
+	// function can provide false and that will adjust behavior of forwarder
+	Active() bool
+}
+
+type TLSData interface {
+	GetCertificate() string
+	GetPrivateKey() string
+}
+
+type Options struct {
+	Logger   *zerolog.Logger
+	LogLevel string
+}
+
+type LoadBalancer struct {
+	id           string
+	runCtx       context.Context
+	killCtx      context.CancelFunc
+	logger       zerolog.Logger
+	poolMap      map[string]ServicePool
+	forwarderMap map[string]*Forwarder
+	mutex        sync.Mutex
+}
+
+// NewLoadBalancer creates new instance of the load balancer
+// using array of pool configuration. For each pool it is
+// assumed that it has unique port, otherwise Listen will
+// fail with the error
+// TODO: Add checkup in constructor for enforce unique port per pool configuration
+func NewLoadBalancer(ctx context.Context, cfgPool []ServicePool, opt Options) (*LoadBalancer, error) {
+	id, err := uuid.NewRandom()
+	if err != nil {
+		return nil, fmt.Errorf("gen id failed")
```

Comment:
Nit: error swallowed rather than wrapped will make this harder to diagnose if there is a fault.